core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotReplicationManager.java

```diff
@@ -0,0 +1,503 @@
+import org.apache.ratis.util.MD5FileUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Class for managing snapshot replication among masters.
+ */
+public class SnapshotReplicationManager {
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotReplicationManager.class);
+  private static final long SNAPSHOT_REQUEST_TIMEOUT_MS =
+      ServerConfiguration.getMs(PropertyKey.MASTER_EMBEDDED_JOURNAL_TRANSPORT_REQUEST_TIMEOUT_MS);
+
+  private final SimpleStateMachineStorage mStorage;
+  private final RaftJournalSystem mJournalSystem;
+  private volatile long mSnapshotRequestTime = 0L;
+  private volatile RaftJournalServiceClient mJournalServiceClient;
+  private volatile SnapshotInfo mDownloadedSnapshot;
+
+  private enum DownloadState {
+    /** No snapshot download is in progress. */
+    IDLE,
+
+    /** Snapshot information is requested from available followers. */
+    REQUEST_INFO,
+
+    /** The latest snapshot data is requested from one of the followers. */
+    REQUEST_DATA,
+
+    /** The latest snapshot is being downloaded from one of the followers. */
+    STREAM_DATA,
+
+    /** A snapshot is downloaded. */
+    DOWNLOADED,
+
+    /** A snapshot is being installed to the journal storage. */
+    INSTALLING,
+  }
+
+  private final AtomicReference<DownloadState> mDownloadState =
+      new AtomicReference<>(DownloadState.IDLE);
+
+  /**
+   * @param journalSystem the raft journal system
+   * @param storage the snapshot storage
+   */
+  public SnapshotReplicationManager(RaftJournalSystem journalSystem,
+      SimpleStateMachineStorage storage) {
+    mStorage = storage;
+    mJournalSystem = journalSystem;
+  }
+
+  /**
+   * @param journalSystem the raft journal system
+   * @param storage the snapshot storage
+   */
+  @VisibleForTesting
+  SnapshotReplicationManager(RaftJournalSystem journalSystem,
+      SimpleStateMachineStorage storage, RaftJournalServiceClient client) {
+    this(journalSystem, storage);
+    mJournalServiceClient = client;
+  }
+
+  /**
+   * Downloads and installs a snapshot from the leader.
+   *
+   * @return a future with the term index of the installed snapshot
+   */
+  public CompletableFuture<TermIndex> installSnapshotFromLeader() {
+    if (mJournalSystem.isLeader()) {
+      return RaftJournalUtils.completeExceptionally(
+          new IllegalStateException("Abort snapshot installation after becoming a leader"));
+    }
+    if (!transitionState(DownloadState.IDLE, DownloadState.STREAM_DATA)) {
+      return RaftJournalUtils.completeExceptionally(
+          new IllegalStateException("Illegal state while installing a snapshot"));
```

Comment:
This exception message will be logged by Ratis. It will only show up in master logs. `transitionState()` will log the expected state and actual state when the expectation is not met.

Reply:
Updated the message to be more descriptive.